enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,543 @@
+apiVersion: foo.example.com/v1
+kind: Foo
+metadata:
+  name: foo-example
+  namespace: default
+spec:
+  ...
+  ...
+  ...
+status:
+  conditions:
+  - type: Migrating # Condition 1
+    status: true
+    reason: "CriticalProcess"
+    message: "Migrating old database schema to new format"
+  - type: UnhealthyDatabase # Condition 2
+    status: true
+    reason: "LivenessProbeFailure"
+    message: "The foo-database failed its liveness probe."
+  - type: BadConnectivity # Condition 3
+    status: true
+    reason: "404"
+    message: "Unable to reach the database"
+  - type: SomethingUnrelated # Condition 4
+    status: true
+    reason: "Unmatched"
+    message: "OLM doesn't care about this condition"
+```
+
+When SPO registers this CR, the Probe CR will be updated like so:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0 # Same name as the CSV.
+  namespace: operators
+spec:
+  specOverride: crdAnnotations # When set to `crdAnnotations` OLM will override any changes to the probe's spec to match those defined via annotations on the CRDs described above.
+  probeResources:
+  - name: foo.example.com
+    upgradeable: "!Migrating"
+    important: "BadConnectivity || UnhealthyDatabase"
+status:
+  conditions:
+  - type: Upgradeable # Overall Upgradeable condition
+    status: false
+    reason: "NotUpgradeable"
+    message: "The operator has communicated that the operator is not upgradeable"
+  probeResources:
+  - kind: foos
+    name: "foo-example"
+    namespace: default
+    conditions:
+    - type: Migrating
+      status: True
+      reason: "CriticalProcess"
+      message: "Migrating old database schema to new format"
+    - type: UnhealthyDatabase # UnhealthyDatabase is an "Important" condition.
+      status: true
+      reason: "LivenessProbeFailure"
+      message: "The foo-database failed its liveness probe."
+    - type: BadConnectivity # BadConnectivity is an "Important" condition.
+      status: true
+      reason: "404"
+      message: "Unable to reach the database"
+```
+
+>Note: `status.probeResources` will be an array of [Rich References](https://github.com/operator-framework/api/blob/bdbc43301d4385ae137472c080d2f4037fbff2b1/pkg/operators/v2alpha1/operator_types.go#L65-L73).
+
+From the presence of the false `Upgradeable` condition, OLM will understand that it should not upgrade the operator. The Conditions used to evaluate the overall `Upgradeable` status of the Operator are stored within the `status.probeResources` array. Notice that the Condition that did not map to an "OLM Supported Condition" was not tracked in the `Probe`.
+
+>Note: If a single CR communicates that the operator is not upgradeable, OLM will not upgrade the operator.
+
+The complexity of communicating state of an operator can be reduced by having the operator introduce a new CRD which is used to communicate its conditions. This creates a single source of truth for the state of the operator.
```

Comment:
It seems that this would be `I can write to the Probe resource myself` option that we discussed but needs to be discussed and added to this enhancement IMO.

Reply:
I'll make sure to call that out at the end of the enhancement.

Reply:
I think this is talking about reducing the cardinality to a singleton CR of a  CRD **owned** by the operator, rather than writing to the `Probe` resource which is provided by OLM.

> but needs to be discussed and added to this enhancement IMO.

Agree that we should mention this as a secondary pattern for communicating to OLM.

Reply:
> I think this is talking about reducing the cardinality to a singleton CR of a CRD owned by the operator, rather than writing to the Probe resource which is provided by OLM.

This snippit does call out the pattern mentioned by @njhale - I have added a small blurb at the end of the proposal highlighting the pattern where the operator writes to the probe resource directly.